modules/nextflow/src/main/groovy/nextflow/trace/AnsiLogObserver.groovy

```diff
@@ -371,34 +386,75 @@ class AnsiLogObserver implements TraceObserver {
     protected String fmtChop(String str, int cols) {
         if( str.size() <= cols )
             return str
-        return cols>3 ? str[0..(cols-3-1)] + '...' : str[0..cols-1]
+        return cols>5 ? str.take(3) + '…' + str.takeRight(cols-1-3) : str[0..cols-1]
     }
 
-    protected String line(ProgressRecord stats) {
+    protected Ansi line(ProgressRecord stats) {
+        final term = ansi()
         final float tot = stats.getTotalCount()
         final float com = stats.getCompletedCount()
         final label = fmtWidth(stats.taskName, labelWidth, Math.max(cols-50, 5))
+        final tagMatch = label =~ / \((.+)\)( *)$/
+        final labelTag = tagMatch ? tagMatch.group(1) : ''
+        final labelSpaces = tagMatch ? tagMatch.group(2) : ''
+        final labelNoTag = label.replaceFirst(/ \(.+\) *$/, "")
+        final labelFinalProcess = labelNoTag.tokenize(':')[-1]
+        final labelNoFinalProcess = labelFinalProcess.length() > 0 ? labelNoTag - ~/$labelFinalProcess$/ : labelNoTag
         final hh = (stats.hash && tot>0 ? stats.hash : '-').padRight(9)
 
-        if( tot == 0  )
-            return "[$hh] process > $label -"
-
         final x = tot ? Math.floor(com / tot * 100f).toInteger() : 0
-        final pct = "[${String.valueOf(x).padLeft(3)}%]".toString()
+        final pct = "${String.valueOf(x).padLeft(3)}%".toString()
+        final numbs = " ${(int)com} of ${(int)tot}".toString()
+
+        // Hash: []
+        term.a(Attribute.INTENSITY_FAINT).a('[').reset()
+        term.fg(Color.BLUE).a(hh).reset()
+        term.a(Attribute.INTENSITY_FAINT).a('] ').reset()
+
+        // Label: process > sayHello
+        if( cols > 180 )
+            term.a(Attribute.INTENSITY_FAINT).a('process > ').reset()
+        term.a(Attribute.INTENSITY_FAINT).a(labelNoFinalProcess).reset()
+        term.a(labelFinalProcess)
+        if( labelTag ){
+            term.fg(Color.YELLOW).a(Attribute.INTENSITY_FAINT).a(' (').reset()
+            term.fg(Color.YELLOW).a(labelTag)
+            term.a(Attribute.INTENSITY_FAINT).a(')').reset().a(labelSpaces)
+        }
+
+        // No tasks
+        if( tot == 0 ) {
+            term.a(' -')
+            return term
+        }
```

Comment:
Well, this code was simple, but now it's totally incomprehensible :cry:

Reply:
Yeah - I spent a while trying to add a helper function into the `term` object, but tied myself into Groovy knots and gave up. So it's very verbose as it stands.

The Jansi library doesn't really give any alternative option though. So a custom helper function was the only thing I could think of.

Reply:
A few comments e.g. what regexs do and similar can help!

Reply:
That I can do!

Reply:
🙏

Reply:
Ok, just pushed a load of comments walking through the code. It was a good exercise for myself as well 😅  Hopefully it's a bit clearer now..

Reply:
Much better now! thanks!